src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/GoToDefinition/RemoteGoToDefinitionService.cs

```diff
@@ -42,12 +43,13 @@ protected override IRemoteGoToDefinitionService CreateService(in ServiceArgs arg
         => RunServiceAsync(
             solutionInfo,
             documentId,
-            context => GetDefinitionAsync(context, position, cancellationToken),
+            context => GetDefinitionAsync(context, position, CreateProjectQueryService(context), cancellationToken),
```

Comment:
Calling the method here is interesting. It freezes the projects/project snapshots to be as at the time of service method call. This might never be a problem, but I don't know. It also always feels clunky to me if we have to put any extra code or logic in this lambda, or this method, other than just plumbing through parameters. I see them as a necessary evil of OOP services, and I really wish they didn't exist. 

I wonder if getting a project query service from a `RemoteDocumentContext` might be worth considering? Either a method, property or extension method would work. It could even go as low as `IDocumentContext` potentionally, though that might be too close to the old "pass around a document snapshot for 'context'" notion that was present before, which would be a valid complaint for sure.

Thoughts?

Reply:
Another way to express this thought: Is it actually possible in cohosting for a "service" to exist that can find projects out of thing air? I don't think so, so I question whether we can/should call it a service and somewhat hide the source of its data.

Reply:
I don't think `RemoteDocumentContext` works because is needed specifically for the scenario where there isn't a document. I think it would weaken `RemoteDocumentContext` significantly if `Snapshot` were nullabile.

Reply:
> It freezes the projects/project snapshots to be as at the time of service method call.

FWIW, I think that's totally expected -- especially since `RemoteDocumentContext` does essentially the same thing. After all, we're supposed to operate on snapshots in co-hosting. Having a service that operates differently (e.g. somehow operates on the current solution snapshot) would be a bad citizen in Roslyn OOP.

Reply:
No worries! I _think_ I've got a better idea. Be prepared to fully review again. :smile:

Reply:
> My thinking is essentially this: I can write `context.Snapshot.WithText("Goo").Project` and I get a project that will not appear in `IProjectQueryService.GetProjects()`.

FWIW, I just took a look at `WithText(...)` and it seems to do a very bad thing. It creates a new `RemoteDocumentSnapshot` for an updated `TextDocument` but passes the old `RemoteProjectSnapshot`. So, `TextDocument.Project` now points to a different project than `RemoteProjectSnapshot._project`! Very naughty! I'm fixing that. :smile:

Reply:
Oh yeah. I fixed that in the `cohost_use_generator` branch, probably should have ported it to main. Turns out it doesn't cause any issues in main though, as luck would have it :D

Reply:
I'm resolving this comment as there's been quite a few changes based on the discussion here. Looking forward to your feedback on the new changes.

Reply:
> Be prepared to fully review again. 😄

Totally not reviewing for the first time 👀